### DISCUSSION TITLE
What is the suggested way to register users for the application.

### DISCUSSION THREAD

**Discussion by Alik2015** [simpleidserver/SimpleIdServer#610] [General]
Given my new application which has a register page.  What is the best way to register users into idserver?  Also would the login by the user be done using idserver views and pages, or is it possible to log the user in using some api call?  If there is any code in the project that can show this would be perfect.

Thanks in advance.

**Comment by simpleidserver** [comment]
Hello @Alik2015 ,

In SimpleIdServer, it is possible to configure one or more `Manual Provisioning Workflows`. These workflows are used by anonymous users to create their local accounts and provision their credentials. 

For example, the URL  `https://openid.simpleidserver.com/master/registration?workflowName=pwd`  is used to create a local account with password authentication.

For more information about `Manual Provisioning Workflow`, please read the documentation : https://simpleidserver.com/docs/iam/manualidentityprovisioning

If you do not wish to use a `Manual Provisioning Workflow`, your website can call the `Users` API to create a local account using the following steps:

1. Configure your client to have access to the `users` scope.
2. Obtain an access token that is valid for the `users` scopes.
3. Execute an HTTP POST request to create a user. The technical identifier of the user must be returned.


```
HTTP REQUEST
Target : https://openid.simpleidserver.com/master/users
Method : POST

Body :
{
"name": "user",
"firstname": "firstname",
"lastname": "lastname",
"email": "email",
"claims": { }
}
```

4. Use the technical identifier of the new user to update the credential as follows:

```
HTTP REQUEST
Target : https://openid.simpleidserver.com/master/users/{id}/credentials
Method : PUT

Body :
{
"type": "pwd",
"value": "PASSWORD"
}
```

**Reply by Alik2015** [reply]
Is it also possible to log user in from the custom app via api sending username and password?  The app will be trusted client on server side.

**Reply by agkoutis** [reply]
Hello @Alik2015,

Technically the Authorization Server is responsible for handling the Authorization Layer of a user. But you can use in the Authentication layer (username & password verification) a microservice (external API). 

I don't know what you are trying to do, but you can check a similar discussion that I have created [here](https://github.com/simpleidserver/SimpleIdServer/discussions/599)

**Reply by Alik2015** [reply]
I agree the auth server should be responsible for signing in would be nice if I could figure out how to implement a Remember Me feature.

**Reply by simpleidserver** [reply]
Hello @Alik2015,

We create a sample project ["https://github.com/simpleidserver/SimpleIdServer/tree/release/4.0.5/samples/CustomUserApi"](https://github.com/simpleidserver/SimpleIdServer/tree/release/4.0.5/samples/CustomUserApi) to demonstrate how to use a REST.API service to authenticate the user via a REST.API.

The `Remember Me` feature is supported by SimpleIdServer [Link](https://github.com/simpleidserver/SimpleIdServer/blob/78f6a5facce17ade5a194be3a6546f834c77a4bc/src/IdServer/SimpleIdServer.IdServer/UI/BaseAuthenticateController.cs#L130) and its implementation cannot be override.

Can-you please give more details about your problem ?

**Reply by Alik2015** [reply]
I was just thinking that remember me should be option as some customers might be using devices they do not own so they don't want to stay signed in and other customers might wish to stay signed in for a long time.

**Reply by agkoutis** [reply]
Are you talking about Single Sign On (user can switch several applications without login) or Cookie Session that never expires (user closes the browser and when re-open Client app doesn't need to login)? Because there is "Remember me" on the login page and it's optionally for the user to enable it.

**Reply by Alik2015** [reply]
I was talking about cookie remember me where user is remembered by the application.  However, I am now thinking using token based sign in where I store the jwt however I want and just have the expiry time extened on the jwt.  I am assuming its possible to extend the jwt expiry?  Reasoning is that jiwt will work across all apps where cookie auth would require more setup like same data protection keys etc.

**Reply by simpleidserver** [reply]
The duration, in `seconds`, of the token / jwt is determined by the property `Client.TokenExpirationTimeInSeconds`. You can directly update this value in the `dbo.Clients` table.

Unfortunately, it is not yet possible to update it via the administration UI. I have created a ticket, #617, to support this feature

**Reply by Alik2015** [reply]
I am using cookie auth and find that when returned in the claims I get nameidentifier set to users name.  My application will link many databases storing data against the user.  For this reason I am thinking that I will need the user's Id instead of the name in other databases because username will be something that can be changed.

For this reason I want to return the id as a claim.  Will I need to modify the source code and customize where the ticket is created or is there an easier way to return the id of the user as a claim.

**Reply by simpleidserver** [reply]
It is possible to set the `sub` claim of the `id_token` like this :

Create a class named `CustomPublicSubjectTypeBuilder` with the following content :

```
public class CustomPublicSubjectTypeBuilder : ISubjectTypeBuilder
{
    public string SubjectType => SUBJECT_TYPE;

    public static string SUBJECT_TYPE = "public";

    public Task<string> Build(HandlerContext context, CancellationToken cancellationToken) => Task.FromResult(context.User.Id);
}
```

Edit the `Program.cs` class and register this class :

```
services.RemoveAll<ISubjectTypeBuilder>();
services.AddTransient<ISubjectTypeBuilder, CustomPublicSubjectTypeBuilder>();
services.AddTransient<ISubjectTypeBuilder, PairWiseSubjectTypeBuidler>();
```

**Reply by Alik2015** [reply]
I thought that email,sub (id) etc come as part of the auth response.  Or is that only if you use jwt that they come as the response?  Also Looking at this code it seems to go in program.cs of idserver project.  The pairWiseSubjectTypeBuilder I am assuming is a class that is internally implemented.

That said, I am guessing the same technique is used to pass out other claims for example the users photo/email etc.

**Reply by simpleidserver** [reply]
Only the JWT token contains the sub. The class CustomPublicSubjectTypeBuilder is utilized to retrieve the public sub, while another one is employed to compute a pairwise sub. 
For more information, you can refer to this documentation: https://connect2id.com/products/server/docs/guides/pairwise-subject-identifiers. 
Therefore, if you wish to replace the sub with the technical identifier, this is the correct place.

For other types of claims, we use Mapping rules to populate them into the id_token. You can find details about the various scope types in the documentation: 
You can see the different kinds of scope in the documentation : https://simpleidserver.com/docs/iam/openid#identity-scope

**Reply by Alik2015** [reply]
As I am converting from OpenIddict the scenario is as follows and I think the above might be a complex way of approaching it.  There might be a simpler way which I am missing.

1) User Registers on the Id Server using the IdServer register page or our custom page and calling api to create account
2) Once account is created, we collect additional data for that user and put it in our database (currently in OpenIddict, we have users id as part of the claim) which we use in our table to link the user data to the account in the id server
3) Many places in our database use the users id because username can be changed, for example custom permissions/policies etc are all stored against the id

Given this my approach which I am not sure will work would be something like the following:
1) Register user as suggested above
2) Once logged in using the server my application will take over by calling the /userinfo endpoint which I am assuming will give us the actual user Id in the db

Once we have that User Id we can do what we need.

As we are converting from OpenIddict and don't which to change too much or rebuild, would be above process give me the User Id that I require?

**Reply by simpleidserver** [reply]
I believe it should work like this :)

In future releases, we will support additional `automatic identity provisioning` workflows to facilitate migrations from OPENIIDDCT/IDSERVER to SimpleIdServer. Currently, we only support 'SCIM' and 'OPENL-LDAP' - you can find more information at https://simpleidserver.com/docs/iam/automaticidentityprovisioning.

**Reply by Alik2015** [reply]
Appreciate that and I know about importing existing users from the details you provided.  For what I describe in my previous comment I just need to get the unique user Id so I can either use method 1 passing it as sub claim or 2 get it from info endpoint.

**Reply by simpleidserver** [reply]
Perhaps a better and simpler solution would be to override the UserRepository and handle both databases. 
I have created a sample project to illustrate how to implement this. What do you think?

Sample project: https://github.com/simpleidserver/SimpleIdServer/tree/release/4.0.5/samples/UseOpenIddictAsDatasource

**Reply by Alik2015** [reply]
Not sure which file/project to look at as there are 2 projects one seems to be the server other seems to be the website.  Also what about my option of creating the user and calling UserInfo endpoint and grabbing the users id from there and just using that.  I am guessing that would also work.  The UserInfo endpoint will always return the logged in users id right?

**Reply by simpleidserver** [reply]
Indeed, your solution can also work, and the userinfo endpoint will always return the information of the authenticated user.

The OpenIddictServer project was pushed by mistake :(
We implemented a [CustomUserRepository](https://github.com/simpleidserver/SimpleIdServer/blob/master/samples/UseOpenIddictAsDatasource/src/UseOpenIddictAsDatasource/Services/CustomUserRepository.cs), which registers the users into the IdentityServer and OpenIddict database.

**Reply by Alik2015** [reply]
@simpleidserver Is email confirmation something I would have to handle on my end or is it implemented as I see a flag for whether or not te email is confirmed?  Also for the following is there a postman sample?

HTTP REQUEST
Target : https://openid.simpleidserver.com/master/users
Method : POST

Body :
{
"name": "user",
"firstname": "firstname",
"lastname": "lastname",
"email": "email",
"claims": { }
}

To create the user using API?

**Reply by simpleidserver** [reply]
I have made some modifications in the `master` branch to provide the option to update the user's property `EmailVerified`.

To create a user, execute the following HTTP request:

```
HTTP POST

URL : https://localhost:5001/master/users

{
    "name": "newName",
    "firstname": "firstname",
    "lastname": "lastname",
    "email": "email",
    "emailverified": true,
    "claims": { }
}
```

To update the `EmailVerified` property, execute the following HTTP request:

```
HTTP PUT
https://localhost:5001/master/users/{{userId}}

{
    "name": "newName",
    "firstname": "firstname",
    "lastname": "lastname",
    "email": "email",
    "emailverified": false,
    "claims": { }
}
```

You can download the Postman project 

[Ticket610Users.postman_collection.json](https://github.com/simpleidserver/SimpleIdServer/files/13590778/Ticket610Users.postman_collection.json)

**Reply by Alik2015** [reply]
Using the postman script it seems does not work.  See below.

![image](https://github.com/simpleidserver/SimpleIdServer/assets/31630746/88364751-6cda-4e64-b2a0-f3619c76e44c)


![image](https://github.com/simpleidserver/SimpleIdServer/assets/31630746/83062d39-840a-4fa7-96bd-c46ea1e75284)


![image](https://github.com/simpleidserver/SimpleIdServer/assets/31630746/479c4fda-3107-40a8-80f9-28f5cb284621)


Am I missing something?

**Reply by simpleidserver** [reply]
Did-you click on the button `Get New Access Token` ?

![image](https://github.com/simpleidserver/SimpleIdServer/assets/10213388/060374e3-59b2-4eb5-89bf-b3c7de67f54c)

**Reply by Alik2015** [reply]
Got it working had to add different url for those who will try you need https://oauth.pstmn.io/v1/callback in the redirect list.

If I wanted to retrieve this by c# I would use M2M instead of web application right?

**Reply by simpleidserver** [reply]
You can use M2M :)

**Reply by Alik2015** [reply]
Thanks sorry going through this bit by bit.  The following does not seem to work

![image](https://github.com/simpleidserver/SimpleIdServer/assets/31630746/8bc6c5b7-4d44-411a-826c-b9fb4e0a50e2)


PUT https://localhost:5001/master/users/cc6ede55-dccc-4c42-8dbb-b00e582470f9/credentials
405
72 ms
JSONError: No data, empty input at 1:1

Body is clearly set.

**Reply by simpleidserver** [reply]
I have created a Postman project to update the user's password.

[Ticket610UpdatePassword.postman_collection.json](https://github.com/simpleidserver/SimpleIdServer/files/13614537/Ticket610UpdatePassword.postman_collection.json)

To accomplish this, add a new password credential and set it as active. Please execute this request:

```
HTTP POST

URL : https://localhost:5001/master/users/3efc07bd-c603-403f-ba47-f3242b12207d/credentials

Body :

{
    "active": true,
    "credential": {
        "type": "pwd",
        "value": "pwd"
    }
}
```

To update a specific credential, please execute this request:

```
HTTP PUT :

URL : https://localhost:5001/master/users/3efc07bd-c603-403f-ba47-f3242b12207d/credentials/{{credentialId}}

Body :

{
    "value": "newpwd"
} 
```

**Comment by Alik2015** [comment]
@simpleidserver appreciate your responses.  I tried using access token for current logged in user to call /master/userinfo endpoint and found that /userinfo does not include the user identifier id.  Would I need to add manually if it is what I need, or is there another way to obtain this.  Also is the docs with example of calling the api's for example /api/userinfo using my token returns 400 invalid token yet /master/userinfo works fine.

**Comment by simpleidserver** [comment]
I made some modifications in the `master` branch to add the possibility of including the `Id` property into a custom claim.

First, add a mapping rule to the `OpenID` scope.

Navigate to the administration website and execute the following tasks:
1. Open the `Scopes` screen and navigate to the `Openid` scope.
2. Select the `Mappers` tab and click on the `Add mapper` button.
3. Choose `Property` and click on `Next`.
4. Fill in the form as follows and click on save:

| Property         | Value  |
| ---------------- | ------ |
| Name             | id     |
| Token Claim name | id     |
| Claim JSON Type  | string |
| User property    | Id     |

5. Open the `Clients` screen and click on the `Add client` button.
6. Select `Web application` and click on the `Next` button.
7. Fill-in the form like this and click on the `Save` button.

| Property            | Value                |
| ------------------- | -------------------- |
| Identifier          | postman              |
| Token Claim name    | password             |
| Name                | postman              |
| Redirection URLS    | http://localhost     |

Finally, import the following Postman project.

[Ticket610.postman_collection.json](https://github.com/simpleidserver/SimpleIdServer/files/13517740/Ticket610.postman_collection.json)

It contains two HTTP requests:
* The first one uses the authorization code grant-type to get an access token.
* The second one calls the `/master/userinfo` endpoint to get the JSON.

This JSON is returned by the userinfo endpoint and contains the technical identifier of the user:

![image](https://github.com/simpleidserver/SimpleIdServer/assets/10213388/e73cd408-2fef-47cb-a095-6312b70efa63)
